File: leetcode.com/1325_delete_leaves_with_a_given_value/solution.c
bool dfs(struct TreeNode *node, int target)
{
    if (node == NULL)
        return false;

    if (node->left == NULL && node->right == NULL)
        return node->val == target;

    if (dfs(node->left, target))
        node->left = NULL;

    if (dfs(node->right, target))
        node->right = NULL;

    return node->left == NULL && node->right == NULL && node->val == target;
}

struct TreeNode *removeLeafNodes(struct TreeNode *root, int target)
{
    if (dfs(root, target))
        return NULL;

    return root;
}

File: leetcode.com/1325_delete_leaves_with_a_given_value/solution.py
class Solution:
    # 13 minutes for solution

    def removeLeafNodes(self, root: TreeNode, target: int) -> TreeNode:

        def dfs(node):
            if not node:
                return

            if node.left is node.right is None:
                return node.val == target

            if dfs(node.left):
                node.left = None

            if dfs(node.right):
                node.right = None

            return node.left is node.right is None and node.val == target

        if dfs(root):
            return None

        return root
